Add skill history, file analysis storage, and top ranked project summaries

Puneet-Maan opened the pull request:
## 📝 Description
Adds the next set of enhancements to Project Insights.

### Summary

Introduces new functionality in `project_insights.py`:

- Storing detailed file analysis for each insight (file counts, sizes, types, largest/newest file)
- Providing a chronological skill history via `list_skill_history`
- Adding summaries for top ranked projects through `summaries_for_top_ranked_projects`
- Normalizing and sorting skills for consistent ordering (including older entries)
- Treating valid but incorrect JSON shapes as corrupted and stashing backups safely

Expands test suite **test_project_insights.py** with coverage for:

- Chronological skill history  
- File analysis fields  
- Summaries for top ranked projects (including score)  
- Non list JSON corruption handling  

Closes: #67 #112 #116

---

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

### Automated

- python3 -m unittest discover -s test -p "test_project_insights.py" -v
- All tests in `test_project_insights.py` now pass, covering:
  - Chronological skill history
  - File analysis persistence
  - Top ranked summaries and score
  - Stashing corrupted/non list JSON
  - All previously existing behaviors

### Manual
- Verified new fields appear correctly in the JSON log
- Confirmed ranking summaries reflect accurate scores and contributor stats
- Checked that corrupted or malformed JSON logs are safely backed up

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

<img width="1118" height="296" alt="image" src="https://github.com/user-attachments/assets/a33ac026-5a01-41d0-933c-32c056fd13c5" />

Weebtrain commented on src/project_insights.py:
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
+    try:
+        return int(value)
+    except Exception:
+        return 0
```
No docstring explaining this method ensures int returned without error

Weebtrain commented on src/project_insights.py:
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
+        try:
+            return datetime.strptime(value, "%Y-%m-%d %H:%M:%S")
+        except Exception:
+            return None
```
No doctring or comments explaining what this does. Good use of safe data conversion.

Weebtrain commented on src/project_insights.py:
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
+            if isinstance(child, dict):
+                stack.append(child)
+
+    return files
```
Useful method. Would serve better as a public method in a utility file for use in other places.

Weebtrain commented on src/project_insights.py:
```diff
@@ -309,9 +414,61 @@ def rank_projects_by_contribution(
+            "skill_count": len(insight.skills),
+        }
+        for insight in list_project_insights(storage_path)
+    ]
```
Docstring is not consistent in listing the return structure, missing "skill_count" mention. No explenation  of parameters.

Puneet-Maan commented on test/test_project_insights.py:
```diff
@@ -9,6 +9,7 @@
 
 import gc
 import json
+import logging
```
I think this is part of the python standard library

Puneet-Maan commented:
@Weebtrain Hey! Thanks for the thorough review. I've pushed updates addressing all your points:
Documentation updates:
- Added full docstrings (with Args/Returns) to _safe_int, _parse_timestamp, _flatten_file_nodes, and _compute_file_analysis
- Expanded docstrings for list_skill_history and summaries_for_top_ranked_projects to list all parameters and returned fields
- Added clearer inline comments in _compute_file_analysis, especially around how the biggest and newest files are chosen
- Explained how storage_path is used inside summaries_for_top_ranked_projects

Test improvements:
- Added comments grouping related assertions and explaining what each group is checking
- Made edge case intentions clearer (like why Beta appears before Alpha in the time based tests)

ThunderIW commented on src/project_insights.py:
```diff
@@ -277,7 +453,15 @@ def record_project_insight(
+
+    Returns:
+        List of ProjectInsight objects sorted by analyzed_at timestamp.
+    """
```
Good function to get the project sorted

ThunderIW reviewed (approved):
Overall, thank you for making the changes @Weebtrain recommended, and good Job on implementing the last bits of our milestone 1 requirements. PR is approved
